File: editors/vscode/src/response.ts
import { RagexError } from "./transport/client";

/**
 * Helpers for interpreting Ragex MCP responses.
 *
 * Every `tools/call` response is wrapped by the server as:
 *   { content: [ { type: "text", text: "<json>" } ] }
 * where `<json>` is the JSON-encoded tool payload. These helpers unwrap that
 * envelope and decode the inner payload.
 */

export interface SearchResult {
  node_type?: string;
  node_id?: string;
  score?: number;
  description?: string;
  context?: {
    file?: string;
    line?: number;
    module?: string;
    function?: string;
    arity?: number;
    callers?: number;
    callees?: number;
  };
}

/** Unwrap an MCP `tools/call` result into the decoded tool payload. */
export function unwrap(result: unknown): unknown {
  if (result === null || result === undefined) {
    return undefined;
  }

  const obj = result as Record<string, unknown>;

  if (obj.error) {
    return obj;
  }

  const content = obj.content;
  if (!Array.isArray(content) || content.length === 0) {
    return result;
  }

  const first = content[0] as Record<string, unknown>;
  const text = first?.text;
  if (typeof text !== "string") {
    return result;
  }

  try {
    return JSON.parse(text);
  } catch {
    return { text };
  }
}

/** Normalize a tool error into a human-readable string. */
export function errorMessage(err: unknown): string {
  if (!err) {
    return "unknown error";
  }
  if (typeof err === "string") {
    return err;
  }
  const e = err as RagexError;
  if (e.message) {
    return e.message;
  }
  return JSON.stringify(err);
}

/** Extract the results array from a search payload. */
export function searchResults(data: unknown): SearchResult[] {
  if (!data || typeof data !== "object") {
    return [];
  }
  const results = (data as Record<string, unknown>).results;
  return Array.isArray(results) ? (results as SearchResult[]) : [];
}

/** Render a search result as a single display line. */
export function formatSearchItem(item: SearchResult): string {
  const id = item.node_id ?? "?";
  const score = typeof item.score === "number" ? item.score.toFixed(3) : "?";
  const desc = item.description ?? "";

  let location = "";
  if (item.context?.file) {
    location = item.context.file;
    if (item.context.line) {
      location = `${location}:${item.context.line}`;
    }
  }

  const parts = [`[${score}] ${id}`];
  if (location) {
    parts.push(location);
  }
  if (desc) {
    parts.push(desc);
  }
  return parts.join("  ");
}

/** Extract a file path and line from a search result, if present. */
export function location(item: SearchResult): { file?: string; line?: number } {
  return { file: item.context?.file, line: item.context?.line };
}

File: editors/nvim/lua/ragex/response.lua
--- Helpers for interpreting Ragex MCP responses.
---
--- Every `tools/call` response is wrapped by the server as:
---
---   { content = { { type = "text", text = "<json>" } } }
---
--- where `<json>` is the JSON-encoded tool payload. These helpers unwrap that
--- envelope and decode the inner payload so callers can work with plain Lua
--- tables.
local M = {}

--- Unwrap an MCP `tools/call` result.
---@param result table|nil
---@return table|nil data, string|nil err
function M.unwrap(result)
  if result == nil then
    return nil, "empty result"
  end

  -- Error shape: { error = {...} } (rare; usually surfaced via JSON-RPC error).
  if result.error then
    return nil, result.error.message or "unknown error"
  end

  local content = result.content
  if type(content) ~= "table" or content[1] == nil then
    -- Some servers return the payload directly.
    return result, nil
  end

  local text = content[1].text
  if type(text) ~= "string" then
    return result, nil
  end

  local ok, decoded = pcall(vim.json.decode, text)
  if ok and decoded ~= nil then
    return decoded, nil
  end

  -- Not JSON (plain text tool output) -- return it verbatim.
  return { text = text }, nil
end

--- Normalize a tool error into a human-readable string.
---@param err table|nil
---@return string
function M.error_message(err)
  if not err then
    return "unknown error"
  end
  if type(err) == "string" then
    return err
  end
  if err.message then
    return err.message
  end
  return vim.inspect(err)
end

--- Best-effort extraction of a list of results from a search payload.
---
--- `semantic_search` / `hybrid_search` return `{ results = [...] }` where each
--- element has `node_id`, `score`, `description`, and optional `context`.
---@param data table|nil
---@return table[] results
function M.search_results(data)
  if type(data) ~= "table" then
    return {}
  end
  return data.results or {}
end

--- Render a search result row as a single display line.
---@param item table
---@return string
function M.format_search_item(item)
  local id = item.node_id or "?"
  local score = item.score and string.format("%.3f", item.score) or "?"
  local desc = item.description or ""

  -- Context (when include_context = true) carries file/line.
  local location = ""
  local ctx = item.context
  if type(ctx) == "table" and ctx.file then
    location = ctx.file
    if ctx.line then
      location = location .. ":" .. tostring(ctx.line)
    end
  end

  local parts = { string.format("[%s] %s", score, id) }
  if location ~= "" then
    table.insert(parts, location)
  end
  if desc ~= "" then
    table.insert(parts, desc)
  end

  return table.concat(parts, "  ")
end

--- Split a `node_id` like "Elixir.My.Module.func/2" into module/function/arity.
---@param node_id string|nil
---@return string|nil module, string|nil func, integer|nil arity
function M.parse_node_id(node_id)
  if type(node_id) ~= "string" then
    return nil, nil, nil
  end

  local module, func, arity = node_id:match("^(.-)%.([%w_!?]+)/(%d+)$")
  if module then
    return module, func, tonumber(arity)
  end

  return node_id, nil, nil
end

--- Extract a file path and line from a search/context payload, if present.
---@param item table
---@return string|nil file, integer|nil line
function M.location(item)
  local ctx = item and item.context
  if type(ctx) == "table" and ctx.file then
    return ctx.file, ctx.line
  end
  return nil, nil
end

return M
